src/engine/sidechain/enginenetworkstream.cpp

```diff
@@ -0,0 +1,184 @@
+    }
+    int copyCount = math_min(writeAvailable, writeRequired);
+    if (copyCount > 0) {
+        (void)m_pOutputFifo->write(buffer, copyCount);
+        // we advance the frame only by the samples we have actually copied
+        // This means in case of buffer full (where we loose some frames)
+        // we do not get out of sync, and the syncing code tries to catch up the
+        // stream by writing silence, once the buffer is free.
+        m_streamFramesWritten += copyCount / m_numOutputChannels;
+    }
+    scheduleWorker();
+}
+
+void EngineNetworkStream::writeSilence(int frames) {
+    //qDebug() << "EngineNetworkStream::writeSilence()" << frames;
+    if (!m_pWorker->threadWaiting()) {
+        // no thread waiting, so we can advance the stream without
+        // buffering
+        m_streamFramesWritten += frames;
+        return;
+    }
+    int writeAvailable = m_pOutputFifo->writeAvailable();
+    int writeRequired = frames * m_numOutputChannels;
+    if (writeAvailable < writeRequired) {
+        qDebug() << "EngineNetworkStream::write() buffer full";
+    }
+    int clearCount = math_min(writeAvailable, writeRequired);
+    if (clearCount > 0) {
+        CSAMPLE* dataPtr1;
+        ring_buffer_size_t size1;
+        CSAMPLE* dataPtr2;
+        ring_buffer_size_t size2;
+        (void)m_pOutputFifo->aquireWriteRegions(clearCount,
+                &dataPtr1, &size1, &dataPtr2, &size2);
+        SampleUtil::clear(dataPtr1,size1);
+        if (size2 > 0) {
+            SampleUtil::clear(dataPtr2,size2);
+        }
+        m_pOutputFifo->releaseWriteRegions(clearCount);
+
+        // we advance the frame only by the samples we have actually cleared
+        m_streamFramesWritten += clearCount / m_numOutputChannels;
+    }
+    scheduleWorker();
+}
+
+void EngineNetworkStream::scheduleWorker() {
+    if (m_pOutputFifo->readAvailable()
+            >= m_numOutputChannels * kNetworkLatencyFrames) {
+        m_pWorker->outputAvailabe();
+    }
+}
+
+void EngineNetworkStream::read(CSAMPLE* buffer, int frames) {
+    int readAvailable = m_pOutputFifo->readAvailable();
+    int readRequired = frames * m_numInputChannels;
+    int copyCount = math_min(readAvailable, readRequired);
+    if (copyCount > 0) {
+        (void)m_pOutputFifo->read(buffer, copyCount);
+        buffer += copyCount;
+    }
+    if (readAvailable < readRequired) {
+        // Fill missing Samples with silence
+        int silenceCount = readRequired - readAvailable;
+        qDebug() << "EngineNetworkStream::write flushed" << readRequired
+                 << "samples";
+        SampleUtil::clear(buffer, silenceCount);
+    }
+}
+
+qint64 EngineNetworkStream::getStreamTimeFrames() {
+    return static_cast<double>(getStreamTimeUs()) * m_sampleRate / 1000000.0;
+}
+
+qint64 EngineNetworkStream::getStreamTimeUs() {
+    return getNetworkTimeUs() - m_streamStartTimeUs;
+}
+
+// static
+qint64 EngineNetworkStream::getNetworkTimeUs() {
```

Comment:
> I think it is intended that libshout uses a clock that is effected by NTP. 

More specifically, libshout's ```shout_sync``` and ```shout_delay``` have a bug. That bug is that they are affected by NTP, clock changes, etc. 

You can see them blaming their users here :tomato: :)
http://lists.xiph.org/pipermail/icecast/2002-February/002262.html

> Only this makes sure that Mixxx and the shoutcast server are synced.

Do you specifically mean "server" here or did you mean libshout? I think you meant libshout as we have a number of servers that implement the protocol that a user might use.

What ```shout_sync``` does is compare the seconds of audio sent out the network socket to the passage of wall-clock time as measured by gettimeofday.  So, it is trying to throttle users of libshout so that they are only sending out audio as fast as their computer's wall clock time is passing.

> Conclusion: we can't use our PerformanceTimer for Windows, since it is QueryPerformanceCounter based which is not NTP calibrated.

Unless I'm missing your point, that's exactly what we want. QElapsedTimer and PerformanceTimer use QueryPerformanceCounter/Frequency so that seems good enough for measuring durations accurately.